aws-janitor/resources/tags.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type Tag struct {
+	Key   string
+	Value string
+}
+
+func NewTag(key *string, value *string) Tag {
+	return Tag{
+		Key:   aws.StringValue(key),
+		Value: aws.StringValue(value),
+	}
+}
+
+// TagMatcher maps keys to valid values. An empty set of values will result in matching tags with any value.
+type TagMatcher map[string]sets.String
+
+// TagMatcherForTags creates a new TagMatcher for the given list of tags provided in key=value format.
+// If "=value" is not provided, then the TagMatcher will match any value for that key.
+// (If the value is empty, only an empty tag value matches.)
+func TagMatcherForTags(tags []string) (TagMatcher, error) {
+	tm := make(TagMatcher)
+	for _, tag := range tags {
+		parts := strings.SplitN(tag, "=", 2)
+		if len(parts) < 1 {
+			return nil, fmt.Errorf("invalid tag: %q", tag)
+		}
+		key := parts[0]
+		if _, ok := tm[key]; !ok {
+			tm[key] = sets.NewString()
+		}
+		if len(parts) == 2 {
+			tm[key].Insert(parts[1])
+		}
+	}
+
+	return tm, nil
+}
+
+func (tm TagMatcher) Matches(tag Tag) bool {
+	vals, ok := tm[tag.Key]
+	if !ok {
+		// No tag matcher for this key
+		return false
+	}
+	if vals.Len() == 0 {
+		// This matcher matches all values for a given key.
+		return true
+	}
+	return vals.Has(tag.Value)
+}
+
+// ManagedPerTags returns whether the given list of tags is matched by all IncludeTags and no ExcludeTags.
+func (opts Options) ManagedPerTags(tags []Tag) bool {
```

Comment:
I don't entirely remember why I opted to use `[]Tag` here, rather than a `map[string]string`, but I think long-term using `map[string]string` for tags instead of `[]Tag` will work out better (for example, it may be useful to pull out `tags[Name]`, which often contains the user-provided name for resources, not available elsewhere.)